pkg/apis/discovery/types.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// EndpointSlice represents a subset of the endpoints that implement a service.
+// For a given service there may be multiple EndpointSlice objects, selected by
+// labels, which must be joined to produce the full set of endpoints.
+type EndpointSlice struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// +optional
+	metav1.ObjectMeta
+	// endpoints is a list of unique endpoint objects in this slice.
+	// +listType=atomic
+	Endpoints []Endpoint
```

Comment:
Do we want to specify a limit to number of Endpoints in a slice?  We never get another chance.  1000?  2000?  5000?

@freehan

Reply:
It will hit the object storage limit naturally. Which means it is around 5000 in normal cases. 

So we might as well set it to 1000?

Reply:
1000 seems reasonable.  We can lift limits if we need to.  Let's define it and validate it.  We have `field.TooLong` but that's not right.  Maybe we want to add `field.TooMany()` ?

Reply:
I added field.TooMany() in my latest commit, hopefully it makes sense.

Reply:
(1000 is also set as the limit here now)